### PR TITLE
feat: Add connection mananger.

### PR DESCRIPTION
**Pull request by kinyoklion**
This adds a connection manager which is used to handle application state and network state and determine the desired connection mode.

### CODE REVIEW

File: packages/shared/sdk-client/src/api/ConnectionMode.ts [hunk 1]
```diff
@@ -9,6 +9,6 @@
  *
  * streaming - The SDK will use a streaming connection to receive updates from LaunchDarkly.
  */
-type ConnectionMode = 'offline' | 'streaming';
```
Comment (kinyoklion) [inline packages/shared/sdk-client/src/api/ConnectionMode.ts:12]: Adding polling as a connection mode. An implementation of polling will be in a later PR.

File: packages/sdk/react-native/src/platform/ConnectionManager.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
+import { ConnectionMode, LDLogger } from '@launchdarkly/js-client-sdk-common';
+
+export enum ApplicationState {
```
Comment (kinyoklion) [inline packages/sdk/react-native/src/platform/ConnectionManager.ts:3]: These "state" enums are internal to the SDK, so I used enums for them. (Versus a union type.)

File: packages/sdk/react-native/src/platform/ConnectionManager.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+export enum NetworkState {
+  /// There is no network available for the SDK to use.
+  Unavailable,
+
+  /// The network is available. Note that network requests may still fail
+  /// for other reasons.
+  Available,
+}
+
+export interface ConnectionDestination {
```
Comment (kinyoklion) [inline packages/sdk/react-native/src/platform/ConnectionManager.ts:23]: Eventually the common client will likely support this interface, but having the abstraction here is nice for testing.

The common interface for the client will likely not expose these as they are for internal control.

File: packages/sdk/react-native/src/platform/ConnectionManager.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  setNetworkAvailability(available: boolean): void;
+  setEventSendingEnabled(enabled: boolean, flush: boolean): void;
+  setConnectionMode(mode: ConnectionMode): Promise<void>;
+  flush(): Promise<void>;
+}
+
+export interface StateDetector {
+  setApplicationStateListener(fn: (state: ApplicationState) => void): void;
```
Comment (kinyoklion) [inline packages/sdk/react-native/src/platform/ConnectionManager.ts:31]: I used basic callback functions for "events" from the state detector.

File: packages/sdk/react-native/src/platform/ConnectionManager.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+  /// The network is available. Note that network requests may still fail
+  /// for other reasons.
+  Available,
+}
+
+export interface ConnectionDestination {
+  setNetworkAvailability(available: boolean): void;
+  setEventSendingEnabled(enabled: boolean, flush: boolean): void;
+  setConnectionMode(mode: ConnectionMode): Promise<void>;
```
Comment (kinyoklion) [inline packages/sdk/react-native/src/platform/ConnectionManager.ts:26]: The current version of setting the connection mode is asynchronous. I am not sure if that is ideal. The process itself is asynchronous, but having a promise that indicates it is "finished" seems risky. Especially lots of independent inputs and without a queue managing transitions.

In the LDClient interface is returns void, but the implementation actually returns a promise and we don't provide any specification of behavior.

File: packages/sdk/react-native/src/platform/ConnectionManager.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  /// connection state will automatically change when network
+  /// availability changes.
+  readonly automaticNetworkHandling: boolean;
+
+  /// Enable handling associated with transitioning between the foreground
+  /// and background.
+  readonly automaticBackgroundHandling: boolean;
+}
+
+export class ConnectionManager {
```
Comment (kinyoklion) [inline packages/sdk/react-native/src/platform/ConnectionManager.ts:55]: This is largely a port of the flutter implementation.

**Review by tanderson-ld — APPROVED** [review]

**Pull request merged**